src/envs/screws.py

```diff
@@ -0,0 +1,509 @@
+
+        # Find which object is the goal screw.
+        assert len(task.goal) == 1
+        goal_pred = list(task.goal)[0]
+        goal_screw = goal_pred.objects[0]
+        assert goal_screw.type.name == 'screw'
+
+        # Draw screws
+        for screw in state.get_objects(self._screw_type):
+            screw_x = state.get(screw, "pose_x")
+            screw_y = state.get(screw, "pose_y")
+            screw_width = state.get(screw, "width")
+            screw_height = state.get(screw, "height")
+            if screw == goal_screw:
+                color = 'gold'
+            else:
+                color = 'grey'
+            rect = plt.Rectangle((screw_x, screw_y),
+                                 screw_width,
+                                 screw_height,
+                                 edgecolor=color,
+                                 facecolor=color,
+                                 label=screw.name)
+            ax.add_patch(rect)
+
+        img = utils.fig2data(fig, dpi=150)
+        plt.close()
+        return [img]
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {
+            self._ScrewPickupable, self._AboveReceptacle, self._HoldingScrew,
+            self._ScrewInReceptacle
+        }
+
+    @property
+    def goal_predicates(self) -> Set[Predicate]:
+        return {self._ScrewInReceptacle}
+
+    @property
+    def types(self) -> Set[Type]:
+        return {self._screw_type, self._gripper_type, self._receptacle_type}
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {
+            self._MoveToScrew, self._MoveToReceptacle, self._MagnetizeGripper,
+            self._DemagnetizeGripper
+        }
+
+    @property
+    def action_space(self) -> Box:
+        # dimensions: [dx, dy, magnetized vs. unmagnetized].
+        lowers = np.array([
+            -(self.rz_x_ub - self.rz_x_lb), -(self.rz_y_ub - self.rz_y_lb), 0.0
+        ],
+                          dtype=np.float32)
+        uppers = np.array([(self.rz_x_ub - self.rz_x_lb),
+                           (self.rz_y_ub - self.rz_y_lb), 1.0],
+                          dtype=np.float32)
+        return Box(lowers, uppers)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        # NOTE: currently, the agent can only pick things up or
+        # drop them iff its staying still. This isn't particularly
+        # realistic...
+        dx, dy, magnetization = action.arr
+        if dx != 0.0 or dy != 0.0:
+            return self._transition_move(state, dx, dy)
+
+        if magnetization > 0.5:
+            return self._transition_magnetize(state)
+        return self._transition_demagnetize(state)
+
+    def _get_tasks(self, num_tasks: int, possible_num_screws: List[int],
+                   rng: np.random.Generator) -> List[Task]:
+        tasks = []
+
+        # Sample center locations for each of the screw clusters.
```

Comment:
I don't care that much about this, but I'm wondering if we could get away without the notion of screw clusters, and just randomly sample positions of all the screws, period?

Reply:
i did that initially, and it led to things generally being grouped further apart. I like the notion of clusters simply because it makes it more likely that we're going to see a lot of different number of screws being held when collecting training data.

Reply:
okay sounds good!